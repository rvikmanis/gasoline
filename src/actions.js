// @flow
import { CHANGE, COMMIT_CHANGES, CANCEL_CHANGES } from './constants'
import * as t from './types'

export function addChange(keyPath: string, subject: mixed) {
  return addChanges({ [keyPath]: subject })
}

export function addChanges(subjectsByKeyPath: t.IDictionary<mixed>) {
  return {
    type: CHANGE,
    payload: {
      subjectsByKeyPath
    }
  }
}

export function commitChanges() {
  return { type: COMMIT_CHANGES }
}

export function cancelChanges() {
  return { type: CANCEL_CHANGES }
}