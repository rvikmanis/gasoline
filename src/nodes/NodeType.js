// @flow
import {
  AbsoluteKeyPathGetter, NodeResolverContext,
  PropResolverContext, writeNodeState,
  Selector, digest
} from '../utils'
import {
  mapValues, get, omit, keys, sortBy, map,
  compose, set, isEqual, reduce, some, identity,
  memoize,
} from 'lodash/fp'
import * as t from '../types'

function resolveProp(context: t.IPropResolverContext) {
  return context.resolveProp()
}

const sortPropResolvers = (propOrder: string[], propResolvers: t.IDictionary<t.IPropResolver>):
  Array<[string, t.IPropResolver]> => compose(
    map(propName => [propName, propResolvers[propName]]),
    sortBy(propName => propOrder.indexOf(propName)),
    keys
  )(propResolvers)

export default function NodeType(typeOptions: t.ITypeOptions = {}): t.INodeCreator {
  const nodeCreator = (instanceOptions = {}): t.INodeBinder => {
    return (bindingContext: t.IBindingContext): t.INode => {
      const {
        keyPath,
        variantKeyPath,
        model = { rootSelector(state: any) { return state } }
      } = bindingContext

      let {
        propResolvers = {},
        propOrder = [],
        defaultProps = {},
        getResolver = (node) => node.resolver,
        getReducer = (node) => node.reducer,
        getEpic = undefined,
        transformInstanceOptions = (options: t.INodeOptions) => options
      } = typeOptions

      let {
        dependencies = {},
        behavior = {},
        ...props
      } = transformInstanceOptions(instanceOptions)

      dependencies = mapValues(AbsoluteKeyPathGetter(keyPath), dependencies)
      props = { ...defaultProps, ...props }

      propResolvers = {
        ...mapValues(() => resolveProp, omit(keys(propResolvers), props)),
        ...propResolvers
      }
      propResolvers = sortPropResolvers(propOrder, propResolvers)

      let node: t.INode;

      const select = Selector(
        digest,
        stateDigest => stateDigest[node.keyPath]
      )

      const getWithState = Selector(
        select,
        nodeState => ({ ...node, state: nodeState })
      )

      let partialNode;

      partialNode = {
        kind: 'node',
        keyPath,
        variantKeyPath,
        dependencies,
        props,
        behavior,
        typeOptions,
        instanceOptions,
        select,
        propResolvers,
        model,
        getWithState,
      }

      function resolver(context: t.INodeResolverContext): t.IDictionary<t.IPropertyValue> {
        const propReducer = (resolvedProps, [propName, propResolver]: [string, t.IPropResolver]) => {
          return {
            ...resolvedProps,
            [propName]: propResolver(
              PropResolverContext(context, propName, resolvedProps)
            )
          }
        }
        return reduce(propReducer, {}, context.node.propResolvers)
      }

      partialNode = {
        ...partialNode,
        resolver: getResolver({ ...partialNode, resolver })
      }

      const reducer = (ctState: t.IControlTreeState, action: t.IAction): t.IControlTreeState => {
        const context = NodeResolverContext(node, ctState, action)
        if (!context.shouldUpdate) {
          return ctState
        }

        const resolvedNodeState = node.resolver(context)
        if (isEqual(resolvedNodeState, context.previousNodeState)) {
          return ctState
        }

        ctState = writeNodeState(ctState, {
          changeSet: (context.isInitial || context.isChange) ? 'next' : 'auto',
          keyPath,
          resolvedNodeState
        })

        return ctState
      }

      node = {
        ...partialNode,
        reducer: getReducer({ ...partialNode, reducer })
      }

      if (typeof getEpic === 'function') {
        node = {
          ...node,
          epic: getEpic(node)
        }
      }

      Object.setPrototypeOf(node, nodeCreator.prototype)
      return node
    }
  }

  nodeCreator.prototype = Object.create(Object.prototype)
  Object.setPrototypeOf(nodeCreator, NodeType.prototype)
  return nodeCreator
}

NodeType.prototype = Object.create(Object.prototype)
