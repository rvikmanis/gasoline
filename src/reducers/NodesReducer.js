// @flow
import * as t from '../types'
import { concatReducers, digest, setWorkingDigest } from '../utils'
import { CHANGE, COMMIT_CHANGES, CANCEL_CHANGES } from '../constants'

function resetAuto(state) {
  let working = { changedKeyPaths: {} }
  Object.keys(state.next).forEach(keyPath => {
    // Triggers dependent nodes to recompute
    working.changedKeyPaths[keyPath] = true
  })
  return { ...state, auto: {}, working }
}

function commit(state) {
  return {
    ...state,
    current: digest(state),
    next: {},
    auto: {},
  }
}

function cancel(state) {
  return {
    ...state,
    current: state.current,
    next: {},
    auto: {}
  }
}

export default function NodesReducer(nodes: t.INode[]) {

  function preNodesReducer(state, action) {
    if (state === undefined) {
      state = {
        next: {},
        current: {},
        auto: {},
        working: {
          isInitial: true
        },
      }
    }

    if (action.type === CHANGE) {
      state = resetAuto(state)
    }

    if (action.type === CHANGE || state.working.isInitial) {
      state = setWorkingDigest(state)
    }

    return state
  }

  function postNodesReducer(state, action) {
    if (action.type === COMMIT_CHANGES || state.working.isInitial) {
      state = commit(state)
    }

    if (action.type === CANCEL_CHANGES) {
      state = cancel(state)
    }

    return { ...state, working: {} }
  }

  return concatReducers(
    preNodesReducer,
    concatReducers(...nodes.map(n => n.reducer)),
    postNodesReducer
  )
}
