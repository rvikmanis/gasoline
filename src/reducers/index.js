export { default as NodesReducer } from './NodesReducer'
