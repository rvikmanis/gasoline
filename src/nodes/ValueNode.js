// @flow
import * as t from '../types'
import NodeType from './NodeType'

export default NodeType({
  defaultProps: {
    defaultValue: null,
    value(dependencies, props) {
      if (props.userValue === undefined) {
        return props.defaultValue
      }
      return props.userValue
    }
  },
  propOrder: ['defaultValue', 'userValue', 'value'],
  propResolvers: {
    userValue: ctx => {
      let result: any

      if (ctx.previousNodeState) {
        result = ctx.previousNodeState.userValue
      }

      if (ctx.isChange) {
        result = ctx.subject
      }

      return result
    },
  }
})
