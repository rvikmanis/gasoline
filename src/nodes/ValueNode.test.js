import NodeType from './NodeType'
import ValueNode from './ValueNode'
import { addChange as change } from '../actions'
import { NodesReducer } from '../reducers'

describe('Value node creator (`ValueNode`)', () => {
  it('is an instance of `NodeType`', () => {
    expect(ValueNode instanceof NodeType).toBe(true)
  })
})

describe('A value node', () => {
  const foo = ValueNode()({ keyPath: 'foo' })

  const bar = ValueNode({
    dependencies: { foo: 'foo' },
    value({ foo }, props) {
      if (!props.userValue) {
        return [foo.value]
      }
      return [props.userValue]
    }
  })({ keyPath: 'bar' })

  const baz = ValueNode({ defaultValue: 'Hello World!' })({ keyPath: 'baz' })

  const updater = NodesReducer([foo, bar, baz])
  let state = updater(undefined, {})

  it('is an instance of `ValueNode`', () => {
    expect(foo instanceof ValueNode).toBe(true)
  })

  it('does initialize as expected', () => {
    expect(state).toMatchSnapshot()
  })

  it('reacts to changes as expected', () => {
    state = updater(state, change('baz', 'BazBazBaz'))
    expect(state).toMatchSnapshot()

    state = updater(state, change('foo', 'Magnificent'))
    expect(state).toMatchSnapshot()

    state = updater(state, change('bar', 'Yohoho'))
    expect(state).toMatchSnapshot()

    state = updater(state, change('bar', undefined))
    expect(state).toMatchSnapshot()

    state = updater(state, change('foo', undefined))
    expect(state).toMatchSnapshot()

    state = updater(state, change('foo', 'Cool'))
    expect(state).toMatchSnapshot()

    expect({
      foo: foo.select(state),
      bar: bar.select(state),
      baz: baz.select(state)
    }).toMatchSnapshot()
  })
})
