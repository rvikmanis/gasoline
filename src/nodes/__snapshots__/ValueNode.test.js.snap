// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A value node does initialize as expected 1`] = `
Object {
  "auto": Object {},
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {},
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 1`] = `
Object {
  "auto": Object {},
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": "BazBazBaz",
      "value": "BazBazBaz",
    },
  },
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 2`] = `
Object {
  "auto": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        "Magnificent",
      ],
    },
  },
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": "BazBazBaz",
      "value": "BazBazBaz",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": "Magnificent",
      "value": "Magnificent",
    },
  },
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 3`] = `
Object {
  "auto": Object {},
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": "Yohoho",
      "value": Array [
        "Yohoho",
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": "BazBazBaz",
      "value": "BazBazBaz",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": "Magnificent",
      "value": "Magnificent",
    },
  },
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 4`] = `
Object {
  "auto": Object {},
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        "Magnificent",
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": "BazBazBaz",
      "value": "BazBazBaz",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": "Magnificent",
      "value": "Magnificent",
    },
  },
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 5`] = `
Object {
  "auto": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
  },
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        "Magnificent",
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": "BazBazBaz",
      "value": "BazBazBaz",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 6`] = `
Object {
  "auto": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        "Cool",
      ],
    },
  },
  "current": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        null,
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": undefined,
      "value": "Hello World!",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": null,
    },
  },
  "next": Object {
    "bar": Object {
      "defaultValue": null,
      "userValue": undefined,
      "value": Array [
        "Magnificent",
      ],
    },
    "baz": Object {
      "defaultValue": "Hello World!",
      "userValue": "BazBazBaz",
      "value": "BazBazBaz",
    },
    "foo": Object {
      "defaultValue": null,
      "userValue": "Cool",
      "value": "Cool",
    },
  },
  "working": Object {},
}
`;

exports[`A value node reacts to changes as expected 7`] = `
Object {
  "bar": Object {
    "defaultValue": null,
    "userValue": undefined,
    "value": Array [
      "Cool",
    ],
  },
  "baz": Object {
    "defaultValue": "Hello World!",
    "userValue": "BazBazBaz",
    "value": "BazBazBaz",
  },
  "foo": Object {
    "defaultValue": null,
    "userValue": "Cool",
    "value": "Cool",
  },
}
`;
